### DISCUSSION TITLE
New PyGlossary StarDict plugin

### DISCUSSION THREAD

**Discussion by ilius** [koreader/koreader#12898] [General]
Users who use PyGlossary to convert a dictionary to StarDict format for purpose of using in KOReader (or `sdcv`) had to use option `merge_syns=True` with [StarDict](https://github.com/ilius/pyglossary/blob/master/doc/p/stardict.md) plugin, because of [issues in `sdcv`](https://github.com/koreader/koreader/issues/5437).

Now this option is removed, and you need to use this new plugin/format instead:
https://github.com/ilius/pyglossary/blob/master/doc/p/stardict_merge_syns.md

If you use a GUI (Gtk, Tkinter or our new web interface), select **StarDict (Merge Syns)** as **Output format**. The same for interactive command line interface.

If you use basic command line interface, add `--write-format=StardictMergeSyns` flag to your command.

Our new [5.0.0](https://github.com/ilius/pyglossary/releases/tag/5.0.0) release is the first tag/release that contains this change.

**Comment by Commodore64user** [comment]
what exactly is this "limitation in `sdcv`"? what's wrong with using `syn` files with koreader?

**Reply by ilius** [reply]
koreader uses sdcv which did not support `.syn` file until version `0.5.2`. I have not tested recent versions and I don't know what version koreader uses.

**Reply by Frenzie** [reply]
0.5.2 since 2017, 0.5.5 (plus a few patches) since 2023.

**Reply by Commodore64user** [reply]
to `.syn` or not to `.syn`, that is the question

https://github.com/Dushistov/sdcv/blob/9c77e910066a471fd37f4c0aa948ff958c836b16/src/stardict_lib.cpp#L839-L864

**Reply by ilius** [reply]
Oh even after 0.5.2 there were problems with it. It's discussed on #5437 and this https://github.com/ilius/pyglossary/pull/250 is the PyGlossary PR that added the option.

**Reply by Commodore64user** [reply]
I've read that issue, but that seems to no longer be the case (the long loading times), as far as I can see, the issue with `syn` files could possibly be related to RAM usage, which for large `syn` files could potentially be quite significant.

**Reply by ilius** [reply]
I think if `syn` is large enough to have RAM issue, it will most likely be really slow, like taking several seconds to show the result. Remember it's not a background-running program. It allocates all that memory when you press enter.

**Reply by Commodore64user** [reply]
Perhaps our resident expert @NiLuJe (or @poire-z) would like to answer the age ol’ question: to `syn` or not to `syn`?

**Reply by poire-z** [reply]
I would if my age ol' memory was of some help.
Only think I could find related to synonym is my comment at https://github.com/koreader/koreader/issues/5437#issuecomment-536322156 (I realized it's been mentionned above). It's possible sdcv has been enhanced since them, dunno.

**Reply by ilius** [reply]
@Commodore64user You can test it with a large Wiktextract (`.jsonl`) file from [kaikki.org](https://kaikki.org/dictionary/index.html). They tend to have many many synonyms / alternate terms. Convert it to StarDict with PyGlossary and test it with `sdcv`.

**Reply by Commodore64user** [reply]
this is as good a time as ever to plug my own https://github.com/Commodore64user/oed_prettifier ;), that has around a million synonyms, and it works totally fine on a kindle voyage, which leads me to question this idea that `syn` files are to be avoided in koreader, but then again, how much memory does building that in-memory map take with 1million entries?

**Reply by Frenzie** [reply]
Simple logic suggests that would take at least a few MB.

But mind this commit from 2020 (the issue referenced above is from 2019): https://github.com/Dushistov/sdcv/commit/994c1c7ae6950b30a60f7c48eb2cbcc72b427620

> This replaces the SynFile::load and SynFile::search with logic borrowed from WordListIndex, resulting in dramatic speed improvement on dictionaries with large synonyms.
> 
> Before:
> 
> ```
> $ time ./sdcv -2 dicts/ hola
> Found 1 items, similar to hola.
> -->Wiktionary (es-en)
> -->hola
> 
> hola (interj)
> 1. hello, hi
> 
> real    0m8.084s
> user    0m7.979s
> sys     0m0.090s
> ```
> 
> After
> 
> ```
> Found 1 items, similar to hola.
> -->Wiktionary (es-en)
> -->hola
> 
> hola (interj)
> 1. hello, hi
> 
> 
> real    0m0.095s
> user    0m0.039s
> sys     0m0.042s
> ```

**Reply by Frenzie** [reply]
PS Only after posting that did I notice the pyglossary `merge_syns` option from October 2020 and the dramatic sdcv performance improvement from December 2020 came from the same author. ;-)

And merged here almost immediately by someone who forgot all about it: https://github.com/koreader/koreader-base/pull/1269

Edit: but guys, it also says fixed right here with all of those references at the bottom of the referenced issue, heh <https://github.com/koreader/koreader/issues/5437#issuecomment-762217555>.

**Reply by Commodore64user** [reply]
> it also says fixed right here

but that is exactly what I have been saying, this hasn't been a problem at all for year thus, why do people keep thinking that syn files are to be avoided with koreader? I am challenging the status quo, in a way. I think is safe to say that, To `syn` or not to `syn`... do whatever you want, It doesn't matter much in the real world, unless you have some ludicrously large amount of syns that would clog the RAM, and important note here, they are cumulative, so if you have 5 dictionaries each with a million synonyms, then you are multiplying your RAM usage, so account for all you files together, not individually. Or use presets ;)

**Reply by ilius** [reply]
Ubuntu 25.04 is the first version that includes [that sdcv fix](https://github.com/Dushistov/sdcv/commit/994c1c7ae6950b30a60f7c48eb2cbcc72b427620). They jumped from 0.5.2 to 0.5.5.
